README.md
# Hexate

[![Build Status](https://travis-ci.org/rjsamson/hexate.svg?branch=master)](https://travis-ci.org/rjsamson/hexate) [![Hex.pm](https://img.shields.io/hexpm/v/hexate.svg)]() [![Hex.pm](https://img.shields.io/hexpm/dt/hexate.svg)]()

A simple module for hexadecimal encoding / decoding in Elixir.

## Note: name change from `hex` to `hexate`

Now that Elixir includes integration with Hex.pm, the library has been
renamed `hexate` to avoid namespace clashes.

This was manifesting itself with intermittent failures when running test
suites. If you experience this, update your deps!

If you really *must* use the old version, it's on the branch `pre-rename-to-hexate`.

## Adding to your `mix.exs`

```elixir
defp deps do
  [
    {:hexate,  ">= 0.6.1"}
  ]
end
```

## Usage

Encode to string (binary):
```elixir
# From a string
iex> Hexate.encode("This is a test.")
     "54686973206973206120746573742e"

# From a char-list
iex> Hexate.encode('This is a test.')
     "54686973206973206120746573742e"

iex> Hexate.encode(123456)
     "1e240"

iex> Hexate.encode(15, 4)
     "000f"

iex> Hexate.encode(15.0, 2)
     "0f"

iex> Hexate.encode(15.0)
     "f"
```

Decode to string (binary):
```elixir
# From a hex string
iex> Hexate.decode("54686973206973206120746573742e")
     "This is a test."

# From a hex char-list
iex> Hexate.decode('54686973206973206120746573742e')
     "This is a test."
```

Encode to hex char-list:
```elixir
# From a unicode char-list
iex> Hexate.encode_to_list('This is a test.')
     '54686973206973206120746573742e'

# From a unicode string
iex> Hexate.encode_to_list("This is a test.")
     '54686973206973206120746573742e'

# From an integer
iex> Hexate.encode_to_list(123456)
     '1e240'
```

Decode to unicode char-list:
```elixir
# From a hex char-list
iex> Hexate.decode_to_list('54686973206973206120746573742e')
     'This is a test.'

# From a hex string
iex> Hexate.decode_to_list("54686973206973206120746573742e")
     'This is a test.'
```

Convert hex to integer:
```elixir
# From hex char-list
iex> Hexate.to_integer('54686973206973206120746573742e')
     438270661302729020147902120434299950

# From hex string
iex> Hexate.to_integer("54686973206973206120746573742e")
     438270661302729020147902120434299950
```

## Contributing

* Fork this repo
* Make a feature branch
* Issue a pull request

## Running tests

```bash
mix deps.get
mix test.watch
```

## Authors

[Robert J Samson](https://github.com/rjsamson)
[David Parry](https://github.com/suranyami)


CHANGELOG.md
# Changelog
**0.6.1**
* Relaxed the Elixir version to be `>= 1.2.0` rather than `~> 1.3` since we're not using anything 1.3-specific. Discovered this while trying to do some work a Raspberry Pi, which still only has 1.2 available at the moment.
**0.6.0**
* Made more consistent use of terms 'char-list' and 'string (binary)'.
* bumped elixir version to 1.0 or greater

**0.5.1**
* Add ex_doc docs
* Version bump for hex.pm publishing

**0.5.0**
* Renamed repo because of possible namespace collision with Hex.pm.

**0.4.1**
* Added encoding with padding digits, handy for generating HTML 2 digit hexate color components.
* Can encode a float now. Useful when converting from CSS floats.
* Fixed some errors in the examples in the README.


test/hex_test.exs
Code.require_file "test_helper.exs", __DIR__

defmodule HexTest do
  use ExUnit.Case

  doctest Hex

  test "encodes an integer to hex string with padding" do
    assert Hex.encode(15, 4) == "000f"
  end

  test "encodes a binary to a hex binary" do
    assert Hex.encode("This is a test.") == "54686973206973206120746573742e"
  end

  test "encodes a list to a hex binary" do
    assert Hex.encode('This is a test.') == "54686973206973206120746573742e"
  end

  test "decodes to a binary from a hex binary" do
    assert Hex.decode("54686973206973206120746573742e") == "This is a test."
  end

  test "decodes to a binary from a hex list" do
    assert Hex.decode('54686973206973206120746573742e') == "This is a test."
  end

  test "encodes a list to a hex list" do
    assert Hex.encode_to_list('This is a test.') == '54686973206973206120746573742e'
  end

  test "encodes a binary to a hex list" do
    assert Hex.encode_to_list("This is a test.") == '54686973206973206120746573742e'
  end

  test "decodes to a list from a hex list" do
    assert Hex.decode_to_list('54686973206973206120746573742e') == 'This is a test.'
  end

 test "decodes to a list from a hex binary" do
    assert Hex.decode_to_list("54686973206973206120746573742e") == 'This is a test.'
  end

  test "converts hex list to an integer" do
    assert Hex.to_integer('54686973206973206120746573742e') == 438270661302729020147902120434299950
  end

  test "converts hex binary to an integer" do
    assert Hex.to_integer("54686973206973206120746573742e") == 438270661302729020147902120434299950
  end

  test "converts an integer to a hex binary" do
    assert Hex.encode(123456) == "1e240"
  end

  test "converts an integer to a hex list" do
    assert Hex.encode_to_list(123456) == '1e240'
  end
end


test/hexate_test.exs
Code.require_file "test_helper.exs", __DIR__

defmodule HexateTest do
  use ExUnit.Case
  doctest Hexate

  test "encodes an integer to hex string with padding" do
    assert Hexate.encode(15, 4) == "000f"
  end

  test "encodes a string to a hex binary" do
    assert Hexate.encode("This is a test.") == "54686973206973206120746573742e"
  end

  test "encodes a char-list to a hex binary" do
    assert Hexate.encode('This is a test.') == "54686973206973206120746573742e"
  end

  test "decodes to a unicode string from a hex string" do
    assert Hexate.decode("54686973206973206120746573742e") == "This is a test."
  end

  test "decodes to a unicode string from a hex char-list" do
    assert Hexate.decode('54686973206973206120746573742e') == "This is a test."
  end

  test "encodes a char-list to a hex char-list" do
    assert Hexate.encode_to_list('This is a test.') ==
      '54686973206973206120746573742e'
  end

  test "encodes a unicode string to a hex char-list" do
    assert Hexate.encode_to_list("This is a test.") ==
      '54686973206973206120746573742e'
  end

  test "decodes to a unicode char-list from a hex char-list" do
    assert Hexate.decode_to_list('54686973206973206120746573742e') ==
      'This is a test.'
  end

 test "decodes to a unicode char-list from a hex string" do
    assert Hexate.decode_to_list("54686973206973206120746573742e") ==
      'This is a test.'
  end

  test "converts hex char-list to an integer" do
    assert Hexate.to_integer('54686973206973206120746573742e') ==
      438270661302729020147902120434299950
  end

  test "converts hex string to an integer" do
    assert Hexate.to_integer("54686973206973206120746573742e") ==
      438270661302729020147902120434299950
  end

  test "converts an integer to a hex string" do
    assert Hexate.encode(123456) == "1e240"
  end

  test "converts an integer to a hex char-list" do
    assert Hexate.encode_to_list(123456) == '1e240'
  end
end


mix.exs
defmodule Hexate.Mixfile do
  use Mix.Project

  def project do
    [ app: :hexate,
      version: "0.6.0",
      elixir: ">= 1.2.0",
      build_embedded: Mix.env == :prod,
      start_permanent: Mix.env == :prod,
      description: description,
      package: package,
      deps: deps ]
  end

  defp description do
    """
    A simple module for working with hex strings in Elixir.
    """
  end

  defp package do
    [
      files: ["lib", "mix.exs", "README*", "LICENSE*", "CHANGELOG*"],
      contributors: ["Rob Samson", "David Parry"],
      licenses: ["MIT"],
      links: %{"GitHub" => "https://github.com/rjsamson/hexate",
              "Docs" => "http://hexdocs.pm/hexate/0.5.1/"}
    ]
  end

  def application do
    [applications: [:logger]]
  end

  defp deps do
    [
      {:mix_test_watch, "~> 0.2", only: :dev},
      {:dogma, "~> 0.1", only: [:dev, :test]},
      {:ex_unit_notifier, "~> 0.1", only: :test},
      {:earmark, "~> 0.1", only: :dev},
      {:ex_doc, "~> 0.11", only: :dev}
    ]
  end
end


lib/hex.ex
defmodule Hexate do
  @moduledoc """
  A simple module to convert to and from hex encoded strings.

  Encodes / decodes both lists and binaries.
  """

  @doc """
  Returns a hex encoded binary from a list, binary or integer.

  ## Examples

      iex> Hexate.encode("This is a test.")
      "54686973206973206120746573742e"

      iex> Hexate.encode('This is a test.')
      "54686973206973206120746573742e"

      iex> Hexate.encode(123456)
      "1e240"

      iex> Hexate.encode(15, 4)
      "000f"

      iex> Hexate.encode(15.0, 2)
      "0f"

      iex> Hexate.encode(15.0)
      "f"
  """
  def encode(int, digits \\ 1)

  def encode(int, digits) when is_integer(int) do
    Integer.to_string(int, 16)
    |> String.downcase
    |> String.rjust(digits, ?0)
  end

  def encode(float, digits) when is_float(float) do
    encode(round(float), digits)
  end

  def encode(str, _digits) when is_binary(str) do
    binary_to_hex_list(str)
    |> IO.iodata_to_binary
  end

  def encode(str, _digits) when is_list(str) do
    list_to_hex(str)
    |> IO.iodata_to_binary
  end

  @doc """
  Returns a hex encoded list from a list, binary or integer.

  ## Examples

      iex> Hexate.encode_to_list("This is a test.")
      '54686973206973206120746573742e'

      iex> Hexate.encode_to_list('This is a test.')
      '54686973206973206120746573742e'

      iex> Hexate.encode_to_list(123456)
      '1e240'
  """
  def encode_to_list(str) when is_binary(str) do
    binary_to_hex_list(str)
  end

  def encode_to_list(str) when is_list(str) do
    list_to_hex(str)
  end

  def encode_to_list(int) when is_integer(int) do
    Integer.to_char_list(int, 16)
    |> :string.to_lower
  end

  @doc """
  Returns a decoded binary from a hex string in either list
  or binary form.

  ## Examples

      iex> Hexate.decode("54686973206973206120746573742e")
      "This is a test."

      iex> Hexate.decode('54686973206973206120746573742e')
      "This is a test."
  """
  def decode(hex_str) when is_binary(hex_str) do
    :binary.bin_to_list(hex_str)
    |> hex_str_to_list
    |> IO.iodata_to_binary
  end

  def decode(hex_str) when is_list(hex_str) do
    hex_str_to_list(hex_str)
    |> IO.iodata_to_binary
  end

  @doc """
  Returns a decoded list from a hex string in either list
  or binary form.

  ## Examples

      iex> Hexate.decode_to_list("54686973206973206120746573742e")
      'This is a test.'

      iex> Hexate.decode_to_list('54686973206973206120746573742e')
      'This is a test.'
  """
  def decode_to_list(hex_str) when is_binary(hex_str) do
    :binary.bin_to_list(hex_str)
    |> hex_str_to_list
  end

  def decode_to_list(hex_str) when is_list(hex_str) do
    hex_str_to_list(hex_str)
  end

  @doc """
  Returns an integer representation of a given string of hex,
  taking a list or a binary as an argument.

  ## Examples

      iex> Hexate.to_integer('54686973206973206120746573742e')
      438270661302729020147902120434299950

      iex> Hexate.to_integer("54686973206973206120746573742e")
      438270661302729020147902120434299950
  """
  def to_integer(hex_str) when is_list(hex_str) do
    List.to_integer(hex_str, 16)
  end

  def to_integer(hex_str) when is_binary(hex_str) do
    String.to_integer(hex_str, 16)
  end

  defp binary_to_hex_list(str) do
    :binary.bin_to_list(str)
    |> list_to_hex
  end

  defp hex_str_to_list([]) do
    []
  end

  defp hex_str_to_list([x,y|tail]) do
    [to_int(x) * 16 + to_int(y) | hex_str_to_list(tail)]
  end

  defp list_to_hex([]) do
    []
  end

  defp list_to_hex([head|tail]) do
    to_hex_str(head) ++ list_to_hex(tail)
  end

  defp to_hex_str(n) when n < 256 do
    [to_hex(div(n, 16)), to_hex(rem(n, 16))]
  end

  defp to_hex(i) when i < 10 do
    0 + i + 48
  end

  defp to_hex(i) when i >= 10 and i < 16 do
    ?a + (i - 10)
  end

  defp to_int(c) when ?0 <= c and c <= ?9 do
    c - ?0
  end

  defp to_int(c) when ?A <= c and c <= ?F do
    c - ?A + 10
  end

  defp to_int(c) when ?a <= c and c <= ?f do
    c - ?a + 10
  end
end


lib/hexate.ex
defmodule Hexate do
  @moduledoc """
  A simple module to convert to and from hex encoded strings.

  Encodes / decodes both char-lists and strings.
  """

  @doc """
  Returns a hex encoded string from a char-list, string or integer.

  ## Examples

      iex> Hexate.encode("This is a test.")
      "54686973206973206120746573742e"

      iex> Hexate.encode('This is a test.')
      "54686973206973206120746573742e"

      iex> Hexate.encode(123456)
      "1e240"

      iex> Hexate.encode(15, 4)
      "000f"

      iex> Hexate.encode(15.0, 2)
      "0f"

      iex> Hexate.encode(15.0)
      "f"
  """
  def encode(int, digits \\ 1)

  def encode(int, digits) when is_integer(int) do
    int
    |> Integer.to_string(16)
    |> String.downcase
    |> String.rjust(digits, ?0)
  end

  def encode(float, digits) when is_float(float) do
    encode(round(float), digits)
  end

  def encode(str, _digits) when is_binary(str) do
    str
    |> binary_to_hex_list
    |> IO.iodata_to_binary
  end

  def encode(str, _digits) when is_list(str) do
    str
    |> list_to_hex
    |> IO.iodata_to_binary
  end

  @doc """
  Returns a hex encoded list from a char-list, string or integer.

  ## Examples

      iex> Hexate.encode_to_list("This is a test.")
      '54686973206973206120746573742e'

      iex> Hexate.encode_to_list('This is a test.')
      '54686973206973206120746573742e'

      iex> Hexate.encode_to_list(123456)
      '1e240'
  """
  def encode_to_list(str) when is_binary(str) do
    binary_to_hex_list(str)
  end

  def encode_to_list(str) when is_list(str) do
    list_to_hex(str)
  end

  def encode_to_list(int) when is_integer(int) do
    int
    |> Integer.to_char_list(16)
    |> :string.to_lower
  end

  @doc """
  Returns a decoded binary from a hex string in either char-list
  or string form.

  ## Examples

      iex> Hexate.decode("54686973206973206120746573742e")
      "This is a test."

      iex> Hexate.decode('54686973206973206120746573742e')
      "This is a test."
  """
  def decode(hex_str) when is_binary(hex_str) do
    hex_str
    |> :binary.bin_to_list
    |> hex_str_to_list
    |> IO.iodata_to_binary
  end

  def decode(hex_str) when is_list(hex_str) do
    hex_str
    |> hex_str_to_list
    |> IO.iodata_to_binary
  end

  @doc """
  Returns a decoded char-list from a hex string in either char-list
  or string form.

  ## Examples

      iex> Hexate.decode_to_list("54686973206973206120746573742e")
      'This is a test.'

      iex> Hexate.decode_to_list('54686973206973206120746573742e')
      'This is a test.'
  """
  def decode_to_list(hex_str) when is_binary(hex_str) do
    hex_str
    |> :binary.bin_to_list
    |> hex_str_to_list
  end

  def decode_to_list(hex_str) when is_list(hex_str) do
    hex_str_to_list(hex_str)
  end

  @doc """
  Returns an integer representation of a given string of hex,
  taking a char-list or a string as an argument.

  ## Examples

      iex> Hexate.to_integer('54686973206973206120746573742e')
      438270661302729020147902120434299950

      iex> Hexate.to_integer("54686973206973206120746573742e")
      438270661302729020147902120434299950
  """
  def to_integer(hex_str) when is_list(hex_str) do
    List.to_integer(hex_str, 16)
  end

  def to_integer(hex_str) when is_binary(hex_str) do
    String.to_integer(hex_str, 16)
  end

  defp binary_to_hex_list(str) do
    str
    |> :binary.bin_to_list
    |> list_to_hex
  end

  defp hex_str_to_list([]) do
    []
  end

  defp hex_str_to_list([x, y | tail]) do
    [to_int(x) * 16 + to_int(y) | hex_str_to_list(tail)]
  end

  defp list_to_hex([]) do
    []
  end

  defp list_to_hex([head | tail]) do
    to_hex_str(head) ++ list_to_hex(tail)
  end

  defp to_hex_str(n) when n < 256 do
    [to_hex(div(n, 16)), to_hex(rem(n, 16))]
  end

  defp to_hex(i) when i < 10 do
    0 + i + 48
  end

  defp to_hex(i) when i >= 10 and i < 16 do
    ?a + (i - 10)
  end

  defp to_int(c) when ?0 <= c and c <= ?9 do
    c - ?0
  end

  defp to_int(c) when ?A <= c and c <= ?F do
    c - ?A + 10
  end

  defp to_int(c) when ?a <= c and c <= ?f do
    c - ?a + 10
  end
end
